### PR TITLE
withLens & reifyLens

### PR DESCRIPTION
**Pull request by aspiwack**
The `withX` family of function is continuation-passing style. Probably
to avoid some allocations. The withLens function was not of this
shape, so I think it deserves another name. I like `reifyLens` for it,
so I used it.

I also added a quick implementation of a `withLens` function in terms
of `reifyLens`. Most certainly we can do a more direct implementation,
but I didn't take the time to think about it.

The getter/setter split of lenses doesn't really make sense for linear
lens. However, there is a presentation using an existential type which
does have such a shape. I decided to use it for `withLens`, by
symmetry with `withIso` and `withPrism`.

Depends on #219 #221 .

### CODE REVIEW

**Comment by aspiwack** [comment]
There is a bit of noise due to the dependency on two different PRs, so either review just the last commit, or wait until #219 is merged.

**Review by utdemir — APPROVED** [review]

File: src/Control/Optics/Linear/Internal.hs
Comment (Divesh-Otwani) [inline src/Control/Optics/Linear/Internal.hs:163]: This is dense but ok.

**Review by Divesh-Otwani — APPROVED** [review]

File: src/Control/Optics/Linear/Internal.hs
Comment (aspiwack) [reply src/Control/Optics/Linear/Internal.hs:163]: There is probably a better definition. One that doesn't indirect through `reifyLens`. I just went for the fastest implementation I could do (though I do find this particular definition illuminating, lenses are a deep and fascinating subject).

**Pull request merged**
